File: src/core/combat/passives/ctx.ts
// 패시브 디스패치 컨텍스트 타입 + 공용 헬퍼.
import type { GameState, Pos, Side, Trigger, Unit } from "../../types.ts";

/** 트리거 발생 컨텍스트(엔진이 훅에서 채워 fireTrigger에 전달). */
export interface TriggerCtx {
  on: Trigger["on"];
  subjectUid?: string; // 이벤트 1차 대상(피격자/턴 주체/상태 대상 등)
  attackerUid?: string; // 가해자/시전자(있으면)
  crit?: boolean;
  damage?: number;
  statusId?: string;
  skillId?: string;
  cell?: Pos;
  winnerSide?: Side; // battleEnd 전용 — 승리 진영
}

/** 한 룰을 평가할 때의 관점: self=룰 소유자 / subject=상대(이벤트 카운터파트) / target=현재 행동 대상. */
export interface RuleCtx {
  owner: Unit;
  subject?: Unit;
  target?: Unit;
  damage?: number; // 흡혈/비율반사용 — 트리거의 피해량(dealtDamage/damaged)
}

export function cmp(a: number, op: "lt" | "lte" | "eq" | "gte" | "gt", b: number): boolean {
  switch (op) {
    case "lt": return a < b;
    case "lte": return a <= b;
    case "eq": return a === b;
    case "gte": return a >= b;
    case "gt": return a > b;
  }
}

/** 그 유닛이 자기 진영 점유 최전열(살아있는 같은 편 최소 열)에 있는가. */
export function isFrontline(state: GameState, u: Unit): boolean {
  let front = Infinity;
  for (const o of state.units) if (o.alive && o.side === u.side) front = Math.min(front, o.pos.col);
  return u.pos.col === front;
}

File: rust/spr-core/src/passives/ctx.rs
//! 패시브 디스패치 컨텍스트 + 공용 헬퍼 (TS `passives/ctx.ts`).
use spr_types::combat::GameState;
use spr_types::data::Pos;

/// 트리거 발생 컨텍스트(엔진이 훅에서 채워 fire_trigger에 전달).
#[derive(Debug, Clone, Default)]
pub struct TriggerCtx {
    pub on: String,
    pub subject_uid: Option<String>,
    pub attacker_uid: Option<String>,
    pub crit: Option<bool>,
    pub damage: Option<i64>,
    pub status_id: Option<String>,
    pub skill_id: Option<String>,
    pub cell: Option<Pos>,
    pub winner_side: Option<String>,
}

impl TriggerCtx {
    pub fn new(on: &str) -> Self {
        TriggerCtx { on: on.to_string(), ..Default::default() }
    }
}

/// 룰 평가 관점: self=소유자 / subject=상대 / target=현재 행동 대상. 인덱스(state.units)로 보유.
#[derive(Debug, Clone, Copy)]
pub struct RuleCtx {
    pub owner: usize,
    pub subject: Option<usize>,
    pub target: Option<usize>,
    pub damage: Option<i64>,
}

impl RuleCtx {
    pub fn of(owner: usize) -> Self {
        RuleCtx { owner, subject: None, target: None, damage: None }
    }
    pub fn with_subject(owner: usize, subject: Option<usize>) -> Self {
        RuleCtx { owner, subject, target: None, damage: None }
    }
}

pub fn cmp(a: f64, op: &str, b: f64) -> bool {
    match op {
        "lt" => a < b,
        "lte" => a <= b,
        "eq" => a == b,
        "gte" => a >= b,
        "gt" => a > b,
        _ => false,
    }
}

/// 정수 비교(부동소수 불요한 조건용 — round/turnCount/col 등).
pub fn cmp_i(a: i64, op: &str, b: i64) -> bool {
    match op {
        "lt" => a < b,
        "lte" => a <= b,
        "eq" => a == b,
        "gte" => a >= b,
        "gt" => a > b,
        _ => false,
    }
}

/// 그 유닛이 자기 진영 점유 최전열(생존 같은 편 최소 열)에 있는가. TS isFrontline.
pub fn is_frontline(state: &GameState, unit_idx: usize) -> bool {
    let u = &state.units[unit_idx];
    let mut front = i64::MAX;
    for o in &state.units {
        if o.alive && o.side == u.side {
            front = front.min(o.pos.col);
        }
    }
    u.pos.col == front
}
